src/hardware/dma.cpp

```diff
@@ -100,29 +106,71 @@ static void perform_dma_io(const DMA_DIRECTION direction,
 	} while (remaining_bytes);
 }
 
-DmaChannel * GetDMAChannel(uint8_t chan) {
-	if (chan<4) {
-		/* channel on first DMA controller */
-		if (DmaControllers[0]) return DmaControllers[0]->GetChannel(chan);
-	} else if (chan<8) {
-		/* channel on second DMA controller */
-		if (DmaControllers[1]) return DmaControllers[1]->GetChannel(chan-4);
+void TANDYSOUND_ShutDown(Section* = nullptr);
+
+static bool activate_primary()
+{
+	assert(!primary);
+	constexpr uint8_t primary_index = 0;
+	primary = std::make_unique<DmaController>(primary_index);
+	return true;
+}
+
+static bool activate_secondary()
+{
+	// We need an AT architecture to support the secondary controller
+	if (!IS_EGAVGA_ARCH) {
+		LOG_WARNING("DMA: Cannot activate the secondary (high) DMA controller "
+		            "because the machine needs to be at least EGA");
+		return false;
 	}
-	return NULL;
+
+	assert(!secondary);
+
+	// The secondary controller and Tandy Sound device conflict in
+	// their use of the 0xc0 IO ports. This is a unique architectural
+	// conflict, so we explicitly shutdown the TandySound device (if
+	// it happens to be running) to meet this request.
+	//
+	TANDYSOUND_ShutDown();
+
+	constexpr uint8_t secondary_index = 1;
+	secondary = std::make_unique<DmaController>(secondary_index);
+	return true;
+}
+
+constexpr uint8_t is_primary(const uint8_t channel_num)
+{
+	return (channel_num < SecondaryMin);
 }
 
-/* remove the second DMA controller (ports are removed automatically) */
-void CloseSecondDMAController(void) {
-	if (DmaControllers[1]) {
-		delete DmaControllers[1];
-		DmaControllers[1]=NULL;
+constexpr uint8_t is_secondary(const uint8_t channel_num)
+{
+	return (channel_num >= SecondaryMin &&
+	        channel_num <= SecondaryMax);
+}
+
+constexpr uint8_t to_secondary_index(const uint8_t channel_num)
+{
+	assert(channel_num >= SecondaryMin);
+	assert(channel_num <= SecondaryMax);
+	return (static_cast<uint8_t>(channel_num - SecondaryMin));
+}
+
+DmaChannel* GetDMAChannel(const uint8_t channel_num)
```

Comment:
Shouldn't this be called `DMA_GetChannel` for consistency? It's not using in many places, so consider renaming it.